django_priority_batch/prioritized_batcher.py

```diff
@@ -0,0 +1,96 @@
+"""Prioritized batcher."""
+import collections
+import logging
+import threading
+
+from django.conf import settings
+
+# Thread-local storage for the global batcher.
+GLOBAL_BATCHER = threading.local()
```

Comment:
I like the idea, but if the batcher crashes, the databases will be out-of-sync and no way to recover.

Reply:
What? This doesn't have to do anything with any database or transactions, this just orders handlers by priority possibly aggregating them and runs them at commit time.

Reply:
I guess we can live with that for a while.

Reply:
No, what I want to say is that this package will never be anything else? It is completely orthogonal to what the handlers will actually do and taking care of things like "crashes" is completely up to the user of this package. There is no generic thing that this package could do.

Reply:
For example, in ES if you wanted to make sure updates are synced:
* In the same transaction as the database update append to a log (a special table in the database) that an update needs to happen, record and bump the item's internal version.
* After database transaction commit, queue an update to ES, possibly using this package to batch updates.
* The handler performs the update and after the update is confirmed to be performed, update the log.
* In case of crashes (or periodically), check the log for any missed updates and commit them.

As you see, ensuring syncness of two databases has nothing to do with this package and would work in the same way with or without it.

Reply:
I'll rephrase. If the thread is killed (this can happen with unhandled exceptions in the batcher or somewhere else, right?), we loose the handlers that were meant to be executed. Maybe half of them were executed and half weren't.

Reply:
I still don't understand your point? If you need stuff to be crash fault tolerant this needs to be handled on another level (e.g., before you queue the handler using this batcher, like inside a database transaction).

But anyway, the thread will not be killed due to unhandled exceptions in the queued handlers. If a handler fails, the exception will only be logged and other handlers will still executed (see code below in `commit`). But of course the process can still crash, be killed or exceptions outside of the batch handlers can interrupt the program after a transaction has been committed - this is why stuff that needs to be crash fault tolerant needs to happen inside a database transaction (e.g., the log that I mentioned above) and not in the batch handler.

Even if the batch handler runs, there could still be network issues preventing the messages from ever reaching their endpoints. But it is not the role of the batcher to make those tolerant to failures (it is likely impossible to solve in a generic fashion anyway as handlers can be vastly different) ;-)